course_info/icommons.py

```diff
@@ -1,80 +1,94 @@
-#from urllib import urlencode
-#from urlparse import urlparse
-#from requests.utils import parse_header_links
-#from django.http import QueryDict
-#from getenv import env
 import drest
-from drest.exc import dRestRequestError, dRestError
+import json
+import logging
+
+from django.core.cache import cache
 from canvas_course_info.settings import aws as settings
 
-import logging
 log = logging.getLogger(__name__)
 
-API_PATH = '/api/course/v2/'
-MAX_ITEMS_PER_REQUEST = 100
-MAX_CONCURRENT_REQUESTS = 4
+CACHE_KEY_COURSE_BY_CANVAS_COURSE_ID = 'course-by-canvas-course-id-{}'
+CACHE_KEY_COURSE_BY_COURSE_INSTANCE_ID = 'course-by-course-instance-id-{}'
+CACHE_KEY_SCHOOL_BY_SCHOOL_ID = 'school-by-school-id-{}'
 
-class CourseUpdateError(dRestRequestError):
-    pass
 
 class ICommonsApi(drest.API):
-
     @classmethod
     def from_request(cls, request):
         icommons_base_url = settings.ICOMMONS_BASE_URL
         access_token = settings.ICOMMONS_API_TOKEN
         return cls(icommons_base_url, access_token)
 
-
     def __init__(self, icommons_base_url, access_token):
         self.icommons_base_url = icommons_base_url
         self.headers = {'Authorization': "Token %s" % access_token,
                         'Content-Type': 'application/json'}
+        API_PATH = settings.ICOMMONS_API_PATH
         super(ICommonsApi, self).__init__(icommons_base_url + API_PATH,
-                                        extra_headers=self.headers,
-                                        timeout=60,
-                                        serialize=True,
-                                        trailing_slash=False)
+                                          extra_headers=self.headers,
+                                          timeout=60,
+                                          serialize=True,
+                                          trailing_slash=False)
 
     def get_course_info(self, course_instance_id):
-        course_info = {}
-        # get the course_instance data
-        try:
-            relative_url = '/course_instances/%s/?format=json' % course_instance_id
-            response = self.make_request('GET', relative_url, headers=self.headers)
-            course_info = response.data.copy()
-        except Exception as e:
-            log.exception(e.message)
+        cache_key = CACHE_KEY_COURSE_BY_COURSE_INSTANCE_ID.format(course_instance_id)
+        course_info = cache.get(cache_key)
+        if course_info is None:
+            course_info = {}
+            # get the course_instance data
+            try:
+                relative_url = '/course_instances/%s/?format=json' % course_instance_id
+                response = self.make_request('GET', relative_url,
+                                             headers=self.headers)
+                course_info = response.data.copy()
```

Comment:
oh, duh, we're returning this.  but nobody passed this struct in, so I'm not entirely sure why we're trying to avoid mangling this struct.  maybe to let the GC clean up the `response` object?  either way, if we really want a full copy, we have to use `copy.deepcopy()`.